File: api-master-data/src/entities/product.entity.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
} from 'typeorm';

@Entity('products')
export class Product {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ name: 'tenant_id', length: 50, nullable: true })
  tenantId: string;

  @Column({ name: 'brand_id', type: 'uuid', nullable: true })
  brandId: string;

  @Column({ name: 'category_id', type: 'uuid', nullable: true })
  categoryId: string;

  @Column({ length: 50, nullable: true })
  code: string;

  @Column({ length: 255 })
  name: string;

  // Tên không dấu để tìm kiếm
  @Column({ name: 'search_name', length: 255, nullable: true })
  searchName: string;

  // Tên viết tắt để tìm kiếm nhanh (VD: "ccdc" cho "Cơm chiên dương châu")
  @Column({ length: 50, nullable: true })
  abbreviation: string;

  @Column({ type: 'text', nullable: true })
  description: string;

  @Column({ type: 'decimal', precision: 15, scale: 2, default: 0 })
  price: number;

  @Column({ name: 'cost_price', type: 'decimal', precision: 15, scale: 2, default: 0, nullable: true })
  costPrice: number;

  @Column({ name: 'image_url', type: 'text', nullable: true })
  imageUrl: string;

  @Column({ length: 50, nullable: true })
  unit: string;

  @Column({ name: 'vat_rate', type: 'decimal', precision: 5, scale: 2, default: 0, nullable: true })
  vatRate: number;

  @Column({ name: 'product_type', length: 50, default: 'single', nullable: true })
  productType: string;

  @Column({ name: 'is_available', type: 'boolean', default: true, nullable: true })
  isAvailable: boolean;

  @Column({ name: 'is_active', type: 'boolean', default: true })
  isActive: boolean;

  @Column({ name: 'sort_order', type: 'int', default: 0 })
  sortOrder: number;

  @Column({ name: 'preparation_time', type: 'int', default: 0, nullable: true })
  preparationTime: number;

  @Column({ name: 'print_dish', type: 'boolean', default: true, nullable: true })
  printDish: boolean;

  @Column({ name: 'print_label', type: 'boolean', default: false, nullable: true })
  printLabel: boolean;

  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date;

  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date;
}

File: ccb-android/app/src/main/java/com/techres/ccb/data/local/entity/ProductEntity.kt
package com.techres.ccb.data.local.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "products",
    indices = [
        Index(value = ["branch_id"]),
        Index(value = ["category_id"]),
        Index(value = ["is_active"]),
        Index(value = ["code"])
    ],
    foreignKeys = [
        ForeignKey(
            entity = CategoryEntity::class,
            parentColumns = ["id"],
            childColumns = ["category_id"],
            onDelete = ForeignKey.SET_NULL
        )
    ]
)
data class ProductEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "branch_id")
    val branchId: String,

    @ColumnInfo(name = "category_id")
    val categoryId: String? = null,

    @ColumnInfo(name = "code")
    val code: String,

    @ColumnInfo(name = "name")
    val name: String,

    // Tên không dấu - dùng để tìm kiếm
    @ColumnInfo(name = "search_name")
    val searchName: String? = null,

    // Tên viết tắt để tìm kiếm nhanh (VD: "ccdc" cho "Cơm chiên dương châu")
    @ColumnInfo(name = "abbreviation")
    val abbreviation: String? = null,

    @ColumnInfo(name = "description")
    val description: String? = null,

    @ColumnInfo(name = "image_url")
    val imageUrl: String? = null,

    @ColumnInfo(name = "price")
    val price: Double,

    @ColumnInfo(name = "cost_price")
    val costPrice: Double = 0.0,

    @ColumnInfo(name = "vat_rate")
    val vatRate: Double = 10.0,

    @ColumnInfo(name = "unit")
    val unit: String? = null,

    @ColumnInfo(name = "type")
    val type: String = "food", // food, drink, other, topping, combo

    @ColumnInfo(name = "is_available")
    val isAvailable: Boolean = true,

    @ColumnInfo(name = "is_active")
    val isActive: Boolean = true,

    @ColumnInfo(name = "sort_order")
    val sortOrder: Int = 0,

    @ColumnInfo(name = "preparation_time")
    val preparationTime: Int = 0, // minutes

    @ColumnInfo(name = "print_to_kitchen")
    val printToKitchen: Boolean = true,

    @ColumnInfo(name = "print_to_bar")
    val printToBar: Boolean = false,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    // Sync fields
    @ColumnInfo(name = "sync_status")
    val syncStatus: String = "synced",

    @ColumnInfo(name = "synced_at")
    val syncedAt: String? = null,

    @ColumnInfo(name = "version")
    val version: Int = 1
)
